apm-sniffer/apm-sdk-plugin/solrj-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/solrj/SolrClientInterceptor.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.skywalking.apm.plugin.solrj.commons.SolrjTags;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.impl.HttpSolrClient;
+import org.apache.solr.client.solrj.request.AbstractUpdateRequest;
+import org.apache.solr.client.solrj.request.QueryRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.common.SolrDocumentList;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.params.UpdateParams;
+import org.apache.solr.common.util.NamedList;
+
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+//
+public class SolrClientInterceptor implements InstanceMethodsAroundInterceptor, InstanceConstructorInterceptor {
+    private static final Pattern URL_REGEX = Pattern.compile("(?<pref>http(s)?://)*(?<domain>[\\w_.\\-\\d]+(:\\d+)?)?/(?<path>solr/(?<collection>[\\w_]+))?(/.*)?");
+
+    @Override
+    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+//      HttpSolrClient(HttpSolrClient.Builder builder)
+        SolrjInstance instance = new SolrjInstance();
+        HttpSolrClient client = (HttpSolrClient) objInst;
+
+        Matcher matcher = URL_REGEX.matcher(client.getBaseURL());
+        if (matcher.find()) {
+            instance.setRemotePeer(matcher.group(3));
+            if (matcher.group(6) != null)
+                instance.setCollection(matcher.group(6));
+        }
+        objInst.setSkyWalkingDynamicField(instance);
+    }
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        SolrRequest<?> request = (SolrRequest<?>) allArguments[0];
+        SolrjInstance instance = (SolrjInstance) objInst.getSkyWalkingDynamicField();
+
+        String collection = instance.getCollection();
+        Object ocollection = allArguments[2];
+        if (ocollection != null) {
+            collection = ocollection.toString();
+        }
+        if (collection == null || "".equals(collection)) {
+            collection = "<empty>";
+        }
+
+        SolrParams params = request.getParams();
+        if (params == null) {
+            params = new ModifiableSolrParams();
+        }
+
+
+        AbstractSpan span = null;
+        if (request instanceof AbstractUpdateRequest) {
+            AbstractUpdateRequest update = (AbstractUpdateRequest) request;
+            String action = "ADD";
+            if (update.getAction() != null) {
+                action = update.getAction().name();
+
+                if (update.getAction() == AbstractUpdateRequest.ACTION.COMMIT) {
```

Comment:
i think update.getAction may exec so many times.

Reply:
@dmsolr Please recheck, if so, then tag is very dangerous, which cost too much memory and network bandwidth.

Reply:
I will review, thanks.